server/src/test/java/org/opensearch/gateway/ReplicaShardBatchAllocatorTests.java

```diff
@@ -0,0 +1,849 @@
+
+    private TestBatchAllocator testBatchAllocator;
+
+    @Before
+    public void buildTestAllocator() {
+        this.testBatchAllocator = new TestBatchAllocator();
+    }
+
+    private void allocateAllUnassignedBatch(final RoutingAllocation allocation) {
+        final RoutingNodes.UnassignedShards.UnassignedIterator iterator = allocation.routingNodes().unassigned().iterator();
+        List<ShardRouting> shardToBatch = new ArrayList<>();
+        while (iterator.hasNext()) {
+            shardToBatch.add(iterator.next());
+        }
+        testBatchAllocator.allocateUnassignedBatch(shardToBatch, allocation);
+    }
+
+    /**
+     * Verifies that when we are still fetching data in an async manner, the replica shard moves to ignore unassigned.
+     */
+    public void testNoAsyncFetchData() {
+        RoutingAllocation allocation = onePrimaryOnNode1And1Replica(yesAllocationDeciders());
+        testBatchAllocator.clean();
+        allocateAllUnassignedBatch(allocation);
+        assertThat(allocation.routingNodes().unassigned().ignored().size(), equalTo(1));
+        assertThat(allocation.routingNodes().unassigned().ignored().get(0).shardId(), equalTo(shardId));
+    }
+
+    /**
+     * Verifies that on index creation, we don't fetch data for any shards, but keep the replica shard unassigned to let
+     * the shard allocator to allocate it. There isn't a copy around to find anyhow.
+     */
+    public void testAsyncFetchWithNoShardOnIndexCreation() {
+        RoutingAllocation allocation = onePrimaryOnNode1And1Replica(
+            yesAllocationDeciders(),
+            Settings.EMPTY,
+            UnassignedInfo.Reason.INDEX_CREATED
+        );
+        testBatchAllocator.clean();
+        allocateAllUnassignedBatch(allocation);
+        assertThat(testBatchAllocator.getFetchDataCalledAndClean(), equalTo(false));
+        assertThat(testBatchAllocator.getShardEligibleFetchDataCountAndClean(), equalTo(0));
+        assertThat(allocation.routingNodes().shardsWithState(ShardRoutingState.UNASSIGNED).size(), equalTo(1));
+        assertThat(allocation.routingNodes().shardsWithState(ShardRoutingState.UNASSIGNED).get(0).shardId(), equalTo(shardId));
+    }
+
+    /**
+     * Verifies that for anything but index creation, fetch data ends up being called, since we need to go and try
+     * and find a better copy for the shard.
+     */
+    public void testAsyncFetchOnAnythingButIndexCreation() {
+        UnassignedInfo.Reason reason = RandomPicks.randomFrom(
+            random(),
+            EnumSet.complementOf(EnumSet.of(UnassignedInfo.Reason.INDEX_CREATED))
+        );
+        RoutingAllocation allocation = onePrimaryOnNode1And1Replica(yesAllocationDeciders(), Settings.EMPTY, reason);
+        testBatchAllocator.clean();
+        allocateAllUnassignedBatch(allocation);
+        assertThat("failed with reason " + reason, testBatchAllocator.getFetchDataCalledAndClean(), equalTo(true));
```

Comment:
why we are expecting failure?

Reply:
This is failure string which is thrown when assert fails. We don't expect failure here.